### ISSUE TITLE
Problem with hot module reloading & createGlobalStyle

### ISSUE THREAD

**Issue by frednomoon** [styled-components/styled-components#2074]
## System:
 - OS: macOS Sierra 10.12.6
 - CPU: x64 Intel(R) Core(TM) i7-4770HQ CPU @ 2.20GHz
 - Memory: 105.95 MB / 16.00 GB
 - Shell: 3.2.57 - /bin/bash
## Binaries:
 - Node: 10.1.0 - ~/.nvm/versions/node/v10.1.0/bin/node
 - Yarn: 1.6.0 - /usr/local/bin/yarn
 - npm: 6.1.0 - ~/.nvm/versions/node/v10.1.0/bin/npm
 - Watchman: 4.7.0 - /usr/local/bin/watchman
## npmPackages:
 - styled-components: ^4.0.0-beta.10 => 4.0.0-beta.10 

Hi there, not sure if I would call this a bug but thought it was worth pointing out.

I have started using the `createGlobalStyle` function in v4, but have run into some issues relating to Hot Module Reloading. Whenever a reload is triggered by something outside the `createGlobalStyle` function, all of its comprising css rules are lost. Here is a simple project which demonstrates this:
https://github.com/frednomoon/styled-hot-reload

It seems from this behaviour that its necessary to reload the module containing the global styles every time the page changes. But, of course, this doesn't happen unless you actually change something inside that module.

I'm guessing its probably possible to solve this with some additional config in my bundler, but obviously would be great to avoid things like that.

**Comment by probablyup** [comment]
Interesting. I'll try and look tonight. There have been a lot of annoying edge cases around HMR lately :/

**Comment by probablyup** [comment]
Ok figured it out. Vanilla HMR replaces the whole file, which is a problem because we keep a counter for cGS to know when to remove global styles when all instances have been removed. If you use `react-hot-loader` in conjunction with parcel's HMR it works correctly.

```
import React from 'react';
import ReactDOM from 'react-dom';
import { hot } from 'react-hot-loader';

import styled, { createGlobalStyle } from 'styled-components';

const GlobalStyle = createGlobalStyle`
  span {
    color: red;
  }
`;

const Container = styled.div`
    display: flex;
    padding: 100px;
    justify-content: center;
    > span {
        font-size: 2rem;
    }
`;

const App = hot(module)(() => (
    <Container>
        <span>THIS SHOULD BE RED</span>
        <GlobalStyle />
    </Container>
));

ReactDOM.render(<App />, document.getElementById('app'));
```

**Comment by frednomoon** [comment]
thanks @probablyup this works great now

**Comment by mxstbr** [comment]
This happens in Next.js 😕 It's pretty annoying especially since I can't just add react-hot-loader I don't think... /cc @timneutkens 

Will reopen this, maybe there's a better way to make `createGlobalStyle` HMR-able from our side?

**Comment by probablyup** [comment]
I think we need to figure out how to refactor away the counting mechanism.
Part of the issue is the old component gets unmounted late during HMR, and
the new component has already rendered, so it sees a count of one due to
the HMR reset and kills the styles.

**Comment by probablyup** [comment]
Give `4.0.0-beta.11.1-hmr2` a shot. It's the change here: https://github.com/styled-components/styled-components/pull/2084

**Comment by vitalybe** [comment]
This still happens to me with `styled-components@4.1.3`

**Comment by deadcoder0904** [comment]
This happens to me too with Gatsby JS with the same version as @vitalybe. Re-opening issue :)

**Comment by gl0gl0** [comment]
This is happening to me as well :(
`styled-components@4.1.1`

**Comment by ConneXNL** [comment]
I am also having an issue with RHR.

```
const Test = styled.div<{ bgColor: string }>`
    background-color: ${({ bgColor }) => bgColor};
`;

Test.defaultProps = {
    bgColor: 'yellow'
};
```

Changing the default props does not re-render the component and it gets stuck no longer updating at all at that point.

However, hot-reloading is working fine in my project for other components where I change the defaultProps. The following reloads just fine:

`const Test = ({ name }: { name: string }) => `Hello ${name}!`;

Test.defaultProps = {
    name: 'Rick'
};`

To me this tells there is something up with Styled-Components.

Any ideas? Could it be the same problem as OP?

**Comment by Sofialp** [comment]
> This is happening to me as well :(, using Razzle
> `styled-components@4.1.1`

**Comment by Sofialp** [comment]
As my collegue @gl0gl0 mention above. We where having this styles bug on localhost, using razzle. 
We where also having this warning on our console that said "Uncaught Error: only one instance of babel-polyfill is allowed". So we delete the import for babel from our hydrate() file on the client side, and this change fixed the styles issue!
Hopefully this can help so users with the same problem. Check that you don't have two babel-polyfill instances being loaded.

**Comment by anthony-langford** [comment]
This is happening with styled-components@4.2.0. Changing the `padding` prop and saving triggers a hot reload, but the padding CSS property stays the same.

```
import React from 'react';
import ReactDOM from 'react-dom';
import styled from 'styled-components';

const Title = styled.h1`
  padding: ${props => props.padding ? props.padding : null};
`;

const App = () => (
  <Title padding="0 40">I am a title</Title>
);

if ((module as any).hot) {
  (module as any).hot.accept();
}

ReactDOM.render(<App />, document.getElementById('app'));
```

**Comment by probablyup** [comment]
You need to supply a unit

**Comment by anthony-langford** [comment]
A unit was supplied in the tested code. The issue remains.

**Comment by BlueAccords** [comment]
Does anyone have any temporary work arounds for this? This is also still happening for me using `styled-components@4.3.0`

**Comment by wa4-fearless-otter** [comment]
Any update on this? Same issue with storybook.

**Comment by gabesu** [comment]
> Any update on this? Same issue with storybook.

Same issue here too with Storybook

**Comment by probablyup** [comment]
Please stop with the "same issue" stuff - provide a fresh reproduction and we'll look into it

**Comment by gabesu** [comment]
> Please stop with the "same issue" stuff - provide a fresh reproduction and we'll look into it

Sorry about that. The issue has been resolved, I just had to upgrade to `styled-components@4.3.2`. Cheers.

**Comment by jdahdah** [comment]
FYI: while it's fixed in `4.3.2`, it seems to persist in `v5.0.0-beta.8`.

**Comment by juanrodriguezarc** [comment]
> FYI: while it's fixed in `4.3.2`, it seems to persist in `v5.0.0-beta.8` (both in conjunction with GatsbyJS).

Same here...  "styled-components": "^5.0.0-beta.8" is working properly, thanks!

**Comment by 3nvi** [comment]
Yup, can verify that it's not working in `v5.0.0-beta.8`.

**Comment by untra** [comment]
styled-components `5.0.0` just launched 🎉 
https://styled-components.com/releases

However this issue is still present in `5.0.0`, with storybook

**Comment by velidan** [comment]
any solution guys or, maybe, workarounds? (Except manual reloading =) )

**Comment by nexusondesign1** [comment]
**Gatsby Fix**

This is the solution to fix that actually provided by gatsby but it's not easy to find it.

After Adding to `gatsby-config.js`
```
resolve: `gatsby-plugin-styled-components`,
      options: {
        // Add any options here
      },
    },
```

and adding to `package.json` as `dependencies ` those fixes the hard reload issue on export.
`"gatsby-plugin-styled-components": "^3.3.0",
  "styled-components": "^5.1.0"
`

also added
`"devDependencies": {
    "babel-plugin-styled-components": "^1.10.7"
  }`

Drop Like if that helps someone that uses `gatsby` and `styled-components` when questions let me know.

**Comment by matiarti** [comment]
Still happening :/

**Comment by Blackxes** [comment]
I did read the whole issue here, but dont quite know if thats the problem i have as well.
Having a styled-component wrapped component passed to hot() reloads the whole tree even components which are next to the styled component rather than within.

I commented createGlobalStyle out but that does not have an effect in any way. Should i make a new issue here?
Or is this related to it?

Greetings, Alex

**Comment by probablyup** [comment]
@frednomoon Could you try out `styled-components@test` and let me know how it goes? (v5)
